### PR TITLE
Implements exit number rewriter, resolves #7

### LINKED ISSUE #7: Re-write exit numbers from highway junction nodes into `junction:ref` way tags
We should re-purpose this project to handle all our tag re-writing. Maybe have a single binary with multiple re-writing passes you can enable or disable on the command line.

Specifically we need to re-write `ref` tag on the `highway=motorway_junction` node into `junction:ref` tag when unambiguously possible.

Context:
- https://github.com/Project-OSRM/osrm-backend/pull/4215
- https://github.com/Project-OSRM/osrm-backend/issues/3400#issuecomment-311326191

cc @willwhite

### PR DESCRIPTION
**Pull request by daniel-j-h**
Re-writes `ref` node tags on `highway=motorway_junction` to `junction:ref` way tags on the adjacent ways when unambiguously possible.

Limitations are documented in the test below: we don't handle the `ref:left` and `ref:right` tags of which there are only a few hundreds on the planet.

In contrast to destination tags we always add a `junction:ref` way tag (except there is already a `junction:ref tag`) but we never check e.g. for `ref` tags already being present on the way. This is because our routing engine _only_ uses `junction:ref` way tags for exit numbers.

https://github.com/mapbox/rewrite-exit-destination-signage/blob/exit-numbers/tests/rewrite.feature

cc @willwhite

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 cmake_minimum_required(VERSION 2.8.11)
-project(exit2destination LANGUAGES C CXX)
+project(osrm-tag-rewriter LANGUAGES C CXX)
```
Comment (MoKob) [inline CMakeLists.txt:2]: all of these project changes look unrelated to the main goal of this PR. Should we maybe split them off into a separate commit?

File: LICENSE [hunk 2]
```diff
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2016 Mapbox
+Copyright (c) 2017 Mapbox
```
Comment (MoKob) [inline LICENSE:3]: also unrelated?

File: src/exits.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+#include <osmium/osm/types.hpp>
+
+#include "util.h"
+
+// Collects node ids and ref=* tags on highway=motorway_junction.
+// http://wiki.openstreetmap.org/wiki/Tag:highway%3Dmotorway_junction
+// http://taginfo.openstreetmap.org/keys/exit_to
+// http://wiki.openstreetmap.org/wiki/Proposed_features/junction_details
+struct ExitsRewriter : osmium::handler::Handler {
+  static const constexpr auto kRefsOnJunctionInPlanet = 91000u;
```
Comment (MoKob) [inline src/exits.h:20]: Is hardcoding numbers of entries in the planet going to be a problem at some point, or is it only for the reservation below?

File: src/exits.h [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+                                    || Equal("primary_link", highway)    //
+                                    || Equal("secondary_link", highway)  //
+                                    || Equal("tertiary_link", highway)); //
+
+    const auto *oneway = way.get_value_by_key("oneway");
+    const auto isOneway = oneway && (Equal("yes", oneway) || //
+                                     Equal("1", oneway) ||   //
+                                     Equal("true", oneway)); //
+
+    const auto isReversed = isOneway && Equal("-1", oneway);
```
Comment (MoKob) [inline src/exits.h]: can this ever be true? From above "-1" would imply `isOneway = false`?

File: src/main.cc [hunk 5]
```diff
@@ -24,20 +25,26 @@ int main(int argc, char **argv) try {
   osmium::io::Reader reader{infile};
   osmium::io::Writer writer{outfile};
 
-  ExitToRewriter xform;
+  DestinationsRewriter destinationsRewriter;
+  ExitsRewriter exitsRewriter;
 
   while (const auto inbuf = reader.read()) {
-    osmium::apply(inbuf, xform);
+    // Todo: refactor as n-stages pipeline with passes enabled/disabled from cmdline
```
Comment (MoKob) [inline src/main.cc:32]: is this ticketed? Could we link to the ticket in the code?

File: tests/rewrite.feature [hunk 6]
```diff
@@ -115,42 +112,44 @@ Feature: Rewrites ExitTo Node Tags to Destination Way tags
...
+            | node | highway           | exit_to | ref |
+            | b    | motorway_junction |       A |   2 |
 
         And the ways
-            | nodes | highway       | oneway | destination:ref |
-            | abd   | motorway      |        |                 |
-            | bc    | motorway_link | yes    | DestinationC    |
+            | nodes | highway       | oneway | destination:ref | junction:ref |
+            | abd   | motorway      |        |                 |              |
+            | bc    | motorway_link | yes    |               B |            3 |
```
Comment (MoKob) [inline tests/rewrite.feature:150]: could you add a test for reversed oneways? From above I would expect them to not be handled correctly.

**Review by MoKob — CHANGES_REQUESTED** [review]
I think there might be a problem with reversed oneways. Also: I am missing some context on how the rewriting works, so I would love if you could give me a short walkthrough of how the process works.

File: src/exits.h (hunk 3 above)
Comment (daniel-j-h) [reply src/exits.h:20]: only reservation

File: src/exits.h (hunk 4 above)
Comment (daniel-j-h) [reply src/exits.h]: urgh good catch

File: src/main.cc (hunk 5 above)
Comment (daniel-j-h) [reply src/main.cc:32]: no there is no ticket for this; it's just something we probably want once we implement more rewriters (?)

more of an idea for the future

**Comment by daniel-j-h** [comment]
👍

The re-writing happens one buffer at a time (roughly 8k osm objects, but there is no guarantee for this): check `src/main.cc` where we read a single buffer from the in file, apply transformations to the buffer, and write it to the out file again. This allows us to re-write while we're streaming the file through our transformations.

What's crucial to understand why this works is in the osm pbf nodes come before ways. Therefore the re-writers can simply store the motorway junction node ids and then once they start to receive ways re-write the ways. We might need to re-structure this pipeline e.g. if we need to re-write way tags to node tags (or do a two pass approach).

The re-writers themselves then work on single buffers. For motorway junction nodes they store the start node for their the way which splits off at the intersection. For all osm objects the re-writers then simply copy over all of their osm tags. For ways they also add the new tag (in case it's not already present).
